source/agora/script/Signature.d

```diff
@@ -0,0 +1,182 @@
+
+    /// Selects behavior of the signature creation & validation algorithm
+    public SigHash sig_hash;
+
+    /***************************************************************************
+
+        Params:
+            signature = the Signature
+            sig_hash = the SigHash
+
+    ***************************************************************************/
+
+    public this (Signature signature, SigHash sig_hash)
+        pure nothrow @safe @nogc
+    {
+        this.signature = signature;
+        this.sig_hash = sig_hash;
+    }
+
+    /***************************************************************************
+
+        todo: unittest
+
+        Returns:
+            The signature + SigHash as a byte array embeddable in a script
+
+    ***************************************************************************/
+
+    public inout(ubyte)[] opSlice () inout pure nothrow @safe /*@nogc*/
+    {
+        return this.signature.toBlob()[] ~ ubyte(this.sig_hash);
+    }
+}
+
+/*******************************************************************************
+
+    Decodes the raw byte representation of a signature to its
+    `Signature` and `SigHash` parts, and validates the `SigHash`
+    to be one of the known flags or accepted combination of flags.
+
+    Params:
+        bytes = contains the <Signature, SigHash> tuple
+        sig_pair = will contain the signature tuple if the Signature was encoded
+                   correctly and the SigHash is one of the known flags or
+                   accepted combination of flags.
+
+    Returns:
+        null if the signature tuple was decoded correctly,
+        otherwise the string explaining the reason why if decoding failed
+
+*******************************************************************************/
+
+public string decodeSignature (const(ubyte)[] bytes,
+    out SigPair sig_pair) pure nothrow @safe @nogc
+{
+    if (bytes.length != SigPair.sizeof)
+        return "Encoded signature tuple is of the wrong size";
+
+    const sig = toSignature(bytes[0 .. Signature.sizeof]);
+    bytes.popFrontN(Signature.sizeof);
+
+    assert(bytes.length == 1);
+    const SigHash sig_hash = cast(SigHash)bytes[0];
+    if (!isValidSigHash(sig_hash))
+        return "Unknown SigHash";
+
+    sig_pair = SigPair(sig, sig_hash);
+    return null;
+}
+
+/*******************************************************************************
+
+    Validates that the given `SigHash` is one of the known flags or one of
+    the accepted combination of flags.
+
+    Params:
+        sig_hash = the `SigHash` to validate
+
+    Returns:
+        true if this is one of the known flags or accepted combination of flags
+
+*******************************************************************************/
+
+private bool isValidSigHash (in SigHash sig_hash) pure nothrow @safe @nogc
+{
+    switch (sig_hash)
+    {
+    case SigHash.All:
+        break;
+
+    default:
+        return false;
+    }
+
+    return true;
+}
```

Comment:
This bothers me, because it means we can have a value that is typed as `SigHash` but is outside the bound of the `enum`. IMO that should not happen to begin with.